### DISCUSSION TITLE
Failing test - FromLowHighUnsigned_x32_Result

### DISCUSSION THREAD

**Discussion by willibrandon** [dotnet/winforms#5716] [General]
﻿I would like to know if this test is failing for anyone else. This test fails for me on all build platforms.

**System.Windows.Forms.Primitives.Tests.Interop.PARAMTests.FromLowHighUnsigned_x32_Result**

[PARAMTests.FromLowHighUnsigned_x32_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L27)

```
System.Windows.Forms.Primitives.Tests.Interop.PARAMTests.FromLowHighUnsigned_x32_Result
   Source: PARAMTests.cs line 27
   Duration: < 1 ms

  Message: 
System.OverflowException : Arithmetic operation resulted in an overflow.

  Stack Trace: 
IntPtr.op_Explicit(IntPtr value)
PARAMTests.FromLowHighUnsigned_x32_Result() line 30
```

I believe the overflow is due to an issue in [Interop.PARAM.FromLowHighUnsigned](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/src/Interop/Interop.PARAM.cs#L22)

```diff
public static IntPtr FromLowHighUnsigned(int low, int high)
     // Convert the int to an uint before converting it to a pointer type,
    // which ensures the high dword being zero for 64-bit pointers.
    // This corresponds to the logic of the MAKELPARAM/MAKEWPARAM/MAKELRESULT
    // macros.
-   => unchecked((nint)(nuint)(uint)ToInt(low, high));
+   => unchecked((nint)(nuint)ToInt(low, high));
```

Since we are now casting to nuint, we no longer need to cast to uint. That was the whole point of using native-sized integers in https://github.com/dotnet/winforms/pull/3603. I believe something was missed there and the original author did not completely understand the change they were making.

Making the suggested change will fix `FromLowHighUnsigned_x32_Result` but will break another PARAM test `FromLowHighUnsigned_x64_Result` because that test is currently expecting an incorrect result for one of its assertions.

[PARAMTests.FromLowHighUnsigned_x64_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L37)

```diff
[ConditionalFact(typeof(ArchitectureDetection), nameof(ArchitectureDetection.Is64bit))]
public unsafe void FromLowHighUnsigned_x64_Result()
{
    Assert.Equal(unchecked((long)0x0000000003040102), (long)PARAM.FromLowHighUnsigned(0x0102, 0x0304));
-    Assert.Equal(unchecked((long)0x00000000F3F4F1F2), (long)PARAM.FromLowHighUnsigned(0xF1F2, 0xF3F4));
+    Assert.Equal(unchecked((long)0xFFFFFFFFF3F4F1F2), (long)PARAM.FromLowHighUnsigned(0xF1F2, 0xF3F4));
}
```

The only thing using `Interop.PARAM.FromLowHighUnsigned()` currently is the TaskDialog to set its progress bar range. Its possible the overflow could be triggered from there somehow.

I would appreciate to hear if the test **System.Windows.Forms.Primitives.Tests.Interop.PARAMTests.FromLowHighUnsigned_x32_Result** has been failing for anyone else and what your thoughts are.

Thank you.

**Comment by weltkante** [comment]
> Since we are now casting to nuint, we no longer need to cast to uint.

This is not correct, did you read the [discussion](https://github.com/dotnet/winforms/pull/3603#discussion_r461491609) we had in the linked PR?

Changing both sign and size of an integer in one operation is ambigous (which operation happens first - sign change or widening? - it produces different results!) and has previously lead to bugs. Obviously the compiler picks one meaning, but even if it happens to match what you intend, being explicit is always better because the reader knows what is meant without looking it up.

Casting first to uint, then widening to nuint, then changing sign again, is explicit in the order of operations and doesn't allow for mistakes.

Now its possible that there is some bug hidden somewhere, normally C# doesn't throw overflow exceptions for arithmetic, but in your stacktrace I'm seeing it being thrown from `IntPtr` code and _not_ from arithmetic, so maybe that exception is hardcoded. Makes me wonder why nobody noticed it though, especially if you claim seeing it happening as test failures. I'll have a look if I can repro.

> Making the suggested change will fix FromLowHighUnsigned_x32_Result but will break another PARAM test FromLowHighUnsigned_x64_Result because that test is currently expecting an incorrect result for one of its assertions.

I'm not sure if its incorrect, I'd have to double check the C/C++ definition, but given what I said above about ambiguity my first guess would be that you seem to have broken the order of operations and now produce incorrect results with your proposed fix?

> I would like to know if this test is failing for anyone else. This test fails for me on all build platforms.

I haven't seen it failing, no. I'll double check though.

**Reply by willibrandon** [reply]
Thank you for your reply.

I did read the [discussion](https://github.com/dotnet/winforms/pull/3603#discussion_r461491609), but my statement regarding no longer needing to cast to uint is different than what you were referring to.

> Casting from `uint` to `nint` changes both sign and size of the value, to the reader its not clear if this does sign extension or not.

I'm not referring to casting from `uint` to `nint`. What I'm saying is that the conversion from `int` to `uint` is unnecessary now that we are using `nuint`.

> I'm not sure if its incorrect, I'd have to double check the C/C++ definition, but given what I said above about ambiguity my first guess would be that you seem to have broken the order of operations and now produce incorrect results with your proposed fix?

Really? Then please tell me why [FromLowHigh_x64_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L23) is expecting `unchecked((long)0xFFFFFFFFF3F4F1F2)` and why [FromLowHighUnsigned_x64_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L37) is expecting `unchecked((long)0x00000000F3F4F1F2)`.

**Reply by weltkante** [reply]
> What I'm saying is that the conversion from int to uint is unnecessary now that we are using nuint.

No it is not, from a readability perspective going from `int` to `nuint` means you are changing both sign and size (on 64 bit platforms) in one operation. If sign change happens first then it will zero-extend, if size change happens first it will sign-extend. As a reader you need to know how the C#/CLR spec defines this operation, otherwise its impossible to understand what it does. Different programming languages define it differently, so easily can cause mistakes.

> Really? Then please tell me why FromLowHigh_x64_Result is expecting unchecked((long)0xFFFFFFFFF3F4F1F2) and why FromLowHighUnsigned_x64_Result is expecting unchecked((long)0x00000000F3F4F1F2).

As far as I can tell without cross comparing with the C/C++ headers, because they are calling different operations. The Windows API uses both variations, some things need to be sign-extended, some need to be zero-extended.

Maybe you're right and I'm missing something, I'm pulling the current github branch right now and will be looking at it in a moment.

**Reply by willibrandon** [reply]
I greatly appreciate you taking time to look at this with me.

If this helps, when I clone the repo, and then run `build.cmd -test`, `Interop.PARAMTests.FromLowHighUnsigned_x32_Result` is skipped according to the TestResults. I see `Interop.PARAMTests.FromLowHighUnsigned_x32_Result` fail when run in Visual Studio.

**Reply by weltkante** [reply]
yes, I've been debugging, VS apparently doesn't respect the `ConditionalFact` attribute and runs the 32 bit test under the 64 bit runtime. In addition, VS doesn't respect its own setting of bitness (I chose running tests as x86 from its UI and it still runs it under the x64 runtime)

Sounds like a VS bug to me and/or a bug in the way the `start-vs.cmd` script sets up VS. Its been a while that I've had to run tests for specific bitness inside VS, it used to work at some point (with VS 2019), I'm not sure when exactly it regressed, if its the switch to VS 2022 or something else.

**Reply by weltkante** [reply]
So to be clear, the exception you're seeing is _not_ coming from the implementation, its coming from the test.

The 32 bit test is doing `(int)PARAM.FromLowHighUnsigned(...)` which is only valid when run on a 32 bit runtime. On a 64 bit runtime the `IntPtr` has value `0x00000000F3F4F1F2` (a `long`) and is too large to cast to an `int`.

When running on a 32 bit runtime the `IntPtr` actually _is_  an `int` so stores `-0xC0B0E0E` (since the highest bit is sign), no overflow happening. The int-arithmetic itself doesn't throw exceptions in C#, only the IntPtr-to-int cast does (on 64 bit, to detect bugs).

**Reply by weltkante** [reply]
I've tested the `ConditionalFact` attribute and the VS processor architecture selection (in the test UI) outside of WinForms, in a fresh xunit project, and neither work - this is a VS 2022 bug (or two, depending on why ConditionalFact is broken).

**Reply by willibrandon** [reply]
Ok, that makes sense. The test `FromLowHighUnsigned_x32_Result` shouldn't even be running under the 64 bit runtime. Thank you for the explanation, that helps a lot. Any idea how I can run it under the 32 bit runtime?

I still don't quite understand why [FromLowHigh_x64_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L23) is expecting `unchecked((long)0xFFFFFFFFF3F4F1F2)` and why [FromLowHighUnsigned_x64_Result](https://github.com/dotnet/winforms/blob/a7761d15f796db76b306cfbc2597d22b46513e5e/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/PARAMTests.cs#L37) is expecting `unchecked((long)0x00000000F3F4F1F2)` but I will take a closer look and see if I can understand why.

Thank you very much for your help. I understand the time and energy required to answer these questions and I greatly appreciate it.

**Reply by RussKie** [reply]
`IntPtr.ToInt32()` on x86 throws if the pointer is larger than an int
https://github.com/dotnet/runtime/blob/f7b441d6c14cfc60c2d02bdc6a087247806efb7c/src/libraries/System.Private.CoreLib/src/System/IntPtr.cs#L98-L107

**Reply by weltkante** [reply]
> I still don't quite understand why FromLowHigh_x64_Result is expecting unchecked((long)0xFFFFFFFFF3F4F1F2) and why FromLowHighUnsigned_x64_Result is expecting unchecked((long)0x00000000F3F4F1F2)

Its for packing two 16-bit values into a pointer-sized value, which on 64 bit has unused bits which need to be filled some way. Some scenarios expect zero-extension, some expect sign-extension. Example macros from the win32 header are `POINTTOPOINTS` which sign-extends when used as `LPARAM` in a window message, and `MAKELPARAM` which zero-extends when used as `LPARAM`.

`WM_NCHITTEST` for example uses a point packed into its `LPARAM` so you'd see sign-extension when manually sending a `WM_NCHITTEST` (as can be seen by the caller of `FromLowHigh`).

`TDM_SET_PROGRESS_BAR_RANGE` instead just wants two values packed (not a point) so you'd be using `MAKELPARAM` in native code, which zero-extends and translates to `FromLowHighUnsigned`

How sensitive code is in regards to these bits being correct is anyones guess, but considering WinForms needs to interact with custom native code (and not just builtin controls) its best to behave the same as Windows does when constructing and sending messages.

**Reply by RussKie** [reply]
> and neither work - this is a VS 2022 bug (or two, depending on why ConditionalFact is broken).

Would you mind raising a VS Feedback for this please?

**Reply by weltkante** [reply]
> Would you mind raising a VS Feedback for this please?

Can do for the bitness selection, but do we know where the `ConditionalFact` attribute is defined? Its pulled in from some Microsoft XUnit extensions via arcade. (When building the repro I just referenced the assembly directly instead of the nuget package because I couldn't figure out where its defined.)

**Reply by RussKie** [reply]
I believe we get it from https://github.com/dotnet/arcade/blob/e7ede87875f41a9b3df898ae08da5ebc96e24f56/src/Microsoft.DotNet.XUnitExtensions/src/Attributes/ConditionalFactAttribute.cs

**Reply by weltkante** [reply]
reported:
* https://developercommunity.visualstudio.com/t/vs-2022-always-runs-tests-as-x64-even-when-selecti/1525842
* https://github.com/dotnet/arcade/issues/7862

**Reply by RussKie** [reply]
Thank you
